File: Solutions/Python/longest_increasing_subsequence.py
"""
    [MEDIUM]
    300. Longest Increasing Subsequence

    Concepts:
    - array
    - dynamic programming

    Stats:
        Runtime | 1755 ms   [Beats 53.27%]
        Memory  | 18.03 MB  [Beats 48.08%]
"""

class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [1]*n

        for i in range(1, n):
            for j in range(i):
                if nums[i] > nums[j]:
                    dp[i] = max(dp[i], dp[j] + 1)

        return max(dp)

File: Solutions/Cpp/longest_increasing_subsequence.cpp
/*
    [MEDIUM]
    300. Longest Increasing Subsequence

    Concepts:
    - array
    - dynamic programming

    Stats:
        Runtime | 43 ms     [Beats 73.48%]
        Memory  | 14.44 MB  [Beats 45.62%]
*/

class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        std::vector<int> dp(nums.size(), 1);

        for (int i = 1; i < nums.size(); i++)
            for (int j = 0; j < i; j++)
                if (nums[i] > nums[j])
                    dp[i] = std::max(dp[i], dp[j] + 1);

        return *(std::max_element(dp.begin(), dp.end()));
    }
};